core/string/src/type.rs

```diff
@@ -0,0 +1,90 @@
+//! Module containing string types public and crate-specific.
+use crate::vtable::SequenceString;
+use crate::{JsStr, JsStringKind};
+use std::alloc::Layout;
+
+pub(crate) mod sealed {
+    use crate::{JsStr, JsStringKind};
+    use std::alloc::Layout;
+
+    /// Seal to prevent others from implementing their own string types.
+    pub trait Sealed {}
+
+    /// Internal trait for crate-specific usage. Contains implementation details
+    /// that should not leak through the API.
+    #[allow(private_interfaces)]
+    pub trait InternalStringType {
```

Comment:
thought: I don't think this needs to be in the private module either.

It can just be a `sealed::Sealed` super trait.

A lot of this tends to come from my experience with icu4x / temporal_rs, so I could be a bit biased on the approach towards sealing.

```rust
mod private {
    pub trait Sealed {}
}

trait TraitOne : private::Sealed { }

trait TraitTwo: private::Sealed { }
```

I think this approach is nice because it's very clear that the trait is private and sealed.